tokenawarebandit.go

```
@@ -0,0 +1,614 @@
+func (t *TokenAwareBanditHostPolicy) resetRing() error {
+	newTR, err := newTokenRing(t.partitioner, t.hostInfoSlice())
+	if err != nil {
+		return err
+	}
+	t.tokenRing = newTR
+
+	for keyspaceName := range t.keyspaces {
+		t.updateReplicas(keyspaceName)
+	}
+
+	return nil
+}
+
+// AddHost adds a single host to the policy after it was discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHost(host *HostInfo) {
+	t.AddHosts([]*HostInfo{host})
+}
+
+// AddHosts adds multiple hosts to the policy after they were discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHosts(hosts []*HostInfo) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	added := false
+
+	for _, host := range hosts {
+		hostID := host.HostID()
+		if hostID == "" {
+			continue
+		}
+		_, ok := t.hosts[hostID]
+		if !ok {
+			t.hosts[hostID] = &tokenAwareBanditHost{
+				hostInfo: host,
+			}
+			added = true
+		}
+	}
+
+	if !added {
+		return
+	}
+
+	err := t.resetRing()
+	if err != nil {
+		t.logger.Println(err)
+	}
+}
+
+// RemoveHost from the policy after it was removed from the cluster.
+func (t *TokenAwareBanditHostPolicy) RemoveHost(host *HostInfo) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	hostID := host.HostID()
+	if hostID == "" {
+		return
+	}
+
+	if _, ok := t.hosts[hostID]; !ok {
+		return
+	}
+
+	delete(t.hosts, hostID)
+	err := t.resetRing()
+	if err != nil {
+		t.logger.Println(err)
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) HostUp(host *HostInfo) {}
+
+func (t *TokenAwareBanditHostPolicy) HostDown(host *HostInfo) {}
+
+// findReplicas gets replicas for the query.
+// If for any reason we can't find replicas, the returned slice's length will be zero.
+// t.mu must be held at least read-only.
+func (t *TokenAwareBanditHostPolicy) findReplicas(qry ExecutableQuery) (token, []*HostInfo) {
+	if qry == nil {
+		return nil, nil
+	}
+
+	routingKey, err := qry.GetRoutingKey()
+	if err != nil || routingKey == nil {
+		return nil, nil
+	}
```

Could we plz document why it' ok to ignore the error and just silently return `nil, nil`?

---

See https://github.com/kiwicom/gocql/pull/9#discussion_r569381795